Data/Text/ICU/Spoof.hsc

```diff
@@ -0,0 +1,273 @@
+
+instance ToBitMask RestrictionLevel where
+  toBitMask ASCII = #const USPOOF_ASCII
+  toBitMask SingleScriptRestrictive = #const USPOOF_SINGLE_SCRIPT_RESTRICTIVE
+  toBitMask HighlyRestrictive = #const USPOOF_HIGHLY_RESTRICTIVE
+  toBitMask ModeratelyRestrictive = #const USPOOF_MODERATELY_RESTRICTIVE
+  toBitMask MinimallyRestrictive = #const USPOOF_MINIMALLY_RESTRICTIVE
+  toBitMask Unrestrictive = #const USPOOF_UNRESTRICTIVE
+
+type URestrictionLevel = Int32
+
+data SpoofCheckResult = CheckOK
+                      | CheckFailed [SpoofCheck]
+                      | CheckFailedWithRestrictionLevel { checks :: [SpoofCheck],
+                                                           level :: RestrictionLevel }
+                deriving (Eq, Show)
+
+data SkeletonTypeOverride = SkeletonSingleScript
+                          | SkeletonAnyCase
+                          deriving (Bounded, Enum, Eq, Show)
+
+instance ToBitMask SkeletonTypeOverride where
+  toBitMask SkeletonSingleScript = #const USPOOF_SINGLE_SCRIPT_CONFUSABLE
+  toBitMask SkeletonAnyCase = #const USPOOF_ANY_CASE
+
+type USkeletonTypeOverride = Int32
+
+makeSpoofCheckResult :: USpoofCheck -> SpoofCheckResult
+makeSpoofCheckResult c =
+  case spoofChecks of
+    [] -> CheckOK
+    _ ->
+      case restrictionLevel of
+        Nothing -> CheckFailed spoofChecks
+        Just l -> CheckFailedWithRestrictionLevel spoofChecks l
+  where spoofChecks = fromBitMask $ fromIntegral $ c .&. #const USPOOF_ALL_CHECKS
+        restrictionValue = c .&. #const USPOOF_RESTRICTION_LEVEL_MASK
+        restrictionLevel = highestValueInBitMask $ fromIntegral $ restrictionValue
+
+-- | Open a spoof checker for checking Unicode strings for lookalike security issues.
+open :: IO MSpoof
+open = wrap =<< handleError uspoof_open
+
+-- | Open a spoof checker previously serialized to bytes using 'serialize'.
+-- The ForeignPtr in the ByteString will remain referenced for the
+-- lifetime of the returned object. The memory it points to must not
+-- be modified until the returned object is closed.
+openFromSerialized :: ByteString -> IO MSpoof
+openFromSerialized b =
+  case toForeignPtr b of
+    (ptr, off, len) -> withForeignPtr ptr $ \p ->
+      wrapWithSerialized ptr =<< handleError (uspoof_openFromSerialized (p `plusPtr` off) (fromIntegral len) nullPtr)
+
+-- | Open a spoof checker given the contents of the "confusables.txt" and "confusablesWholeScript.txt"
+-- files as described in Unicode UAX #39.
+openFromSource :: ByteString -> ByteString -> IO MSpoof
+openFromSource confusables confusablesWholeScript =
+  unsafeUseAsCStringLen confusables $ \(cptr, clen) ->
+    unsafeUseAsCStringLen confusablesWholeScript $ \(wptr, wlen) ->
+      wrap =<< handleError (uspoof_openFromSource cptr (fromIntegral clen) wptr (fromIntegral wlen) nullPtr nullPtr)
+
+-- | Get the checks performed by a spoof checker.
+getChecks :: MSpoof -> IO [SpoofCheck]
+getChecks s = do
```

Comment:
General note about functions in this module: there's no value to using `do` block for a single statement.